netdog: set IPv6 accept-ra for primary interface via config rather than sysctl

zmrow opened the pull request:
**Issue number:**
Related to bottlerocket-os/bottlerocket#3143 

*A separate PR will be opened by @markusboehme for the related patches to `wicked`*

**Description of changes:**
```
    netdog: Add wicked IPv6 config structures to support RA
    
    This change adds the necessary structures to wicked in order to support
    setting the `accept_ra` IPv6 option via interface config.
```
```
    netdog: Set default accept-ra setting via config rather than sysctl
    
    Previously, we set the IPv6 accept-ra setting for the primary interface
    via `systemd-sysctl` during the wicked "install" helper.  Setting this
    sysctl via the helper can cause a race condition between the kernel and
    wicked.  The kernel sets a flag indicating a router advertisement has
    been received (`IF_RA_RCVD`), but only after it completes duplicate
    address detection and decides whether to send a router solicitation.  If
    the sysctl isn't set by the time duplicate address detection completes,
    the solicitation doesn't happen and the `IF_RA_RCVD` flag doesn't get
    set.  wicked uses this flag to decide whether or not to kick off the
    state machine that handles DHCP6.
    
    This change adds the accept-ra setting to the interface config if the
    primary interface is set up via kernel command line.  The primary
    interface is supplied via kernel command line for AWS and VMware
    variants.  This change also removes the accept-ra setting from the
    `systemd-sysctl` config.  Allowing wicked to manage this setting
    eliminates any chance of races between sysctl/kernel/wicked.
```
In the future we will support a network config setting that sets this in config for any interface specified.
**Testing done:**
* All unit tests continue to pass
* Launch 6,000 AWS instances (in batches of 2,000) including these fixes and the aforementioned pending `wicked` patches (thanks @markusboehme! ) into an IPv6 EKS cluster.  Observe all instances join the cluster, indicating they all received an IPv6 address successfully. 
* Also launched an `aws-k8s-1.24` instance into a "normal" IPv4 cluster successfully.
* Launch a `vmware-k8s-1.24` VM and observe it join my cluster successfully.
* Observe the proper config being set for `eth0`:
```
<interface><name>eth0</name>...<ipv6><accept-ra>router</accept-ra></ipv6></interface>
```
* Ovserve the proper sysctls being set via `systemd-sysctl` and from wicked
```
bash-5.1# cat /etc/sysctl.d/90-primary_interface.conf 
-net.ipv4.conf.eth0.rp_filter = 2

bash-5.1# sysctl net.ipv6.conf.eth0.accept_ra
net.ipv6.conf.eth0.accept_ra = 2
```


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

zmrow committed:
netdog: Add wicked IPv6 config structures to support RA

This change adds the necessary structures to wicked in order to support
setting the `accept_ra` IPv6 option via interface config.  Doing this
in config allows wicked to manage this option and avoids any race
conditions that may occur between wicked and the kernel when attempting
to set this via sysctl.

yeazelm commented on sources/api/netdog/test_data/wicked/eno1-ra.xml:
```diff
@@ -0,0 +1 @@
+<interface><name>eno1</name><control><mode>boot</mode><link-detection><require-link></require-link></link-detection></control><ipv4:dhcp><enabled>true</enabled></ipv4:dhcp><ipv6><accept-ra>router</accept-ra></ipv6></interface>
```
Is it correct to do this if the interface is not configured for IPv6 otherwise?

yeazelm replied:
I think from further looking, this probably won't affect anything so I'm fine with it remaining even if it looks a little odd.

zmrow replied:
This feels pretty safe given that it only gets added if the interface is configured via kernel command line, which currently only happens in AWS/VMware variants where DHCP6 is optional by default.  At least currently, it would be a bit of a chore to specify an interface via kernel command line with other variants.

zmrow replied:
Ideally, and I mentioned this in my comments with @markusboehme , we will make this more configurable in the future via net config.

markusboehme commented on sources/api/netdog/src/cli/generate_net_config.rs:
```diff
@@ -48,8 +50,16 @@ pub(crate) fn run() -> Result<()> {
+        // fix some defaults to match legacy behavior.
+        // Note: we only allow 1 interface to be listed via kernel command line, so this will only
+        // be added to a single interface
+        if from_cmd_line {
+            interface.accept_ra();
+        }
```
I don't understand why this is conditional on the interface being configured via the kernel command line. Before this change, the interface's `accept_ra` sysctl would be set to `2` regardless. Should this instead be conditional on it being the primary interface? (With #2615 later it would also need to check for conflicting explicit configuration, but not yet.)

zmrow replied:
Good question.  We want to make this fully configurable in #2615 .  

At the moment, metal deployments via EKS-A don't support IPv6, so limiting this change to variants that get their config via the command line (AWS/VMware) feels pretty safe from any regression.  AWS/VMware variants configure a single interface (the primary) via command line, so this config will only be applied to that primary interface.

stmcginnis reviewed (approved):
Changes look good, and coverage appears thorough enough.

bcressey commented on sources/api/netdog/src/wicked/mod.rs:
```diff
@@ -206,6 +206,12 @@ impl WickedInterface {
         }
     }
 
+    /// Add config to accept IPv6 router advertisements
+    // TODO: expose a network config option for this
+    pub(crate) fn accept_ra(&mut self) {
+        self.ipv6 = Some(WickedIpv6::default())
+    }
```
I would've expected this change, and the changes to `write_config_file` to be in the previous commit.

zmrow replied:
Good catch.... That was how I intended it to be; must have mixed it up somewhere along the way.  Will fix.

bcressey reviewed (approved):
LGTM overall, but if you have a chance to clean up the commits so that all the necessary modifications to `wicked/mod.rs` happen together, that seems like a more logical sequence.

zmrow committed:
netdog: Set default accept-ra setting via config rather than sysctl

Previously, we set the IPv6 accept-ra setting for the primary interface
via `systemd-sysctl` during the wicked install helper.  Setting this
sysctl via the helper can cause a race condition between the kernel and
wicked.  The kernel sets a flag indicating a router advertisement has
been received (`IF_RA_RCVD`), but only after it completes duplicate
address detection and decides whether to send a router solicitation.  If
the sysctl isn't set by the time duplicate address detection completes,
the solicitation doesn't happen and the `IF_RA_RCVD` flag doesn't get
set.  wicked uses this flag to decide whether or not to kick off the
state machine that handles DHCP6.

This change adds the accept-ra setting to the interface config if the
primary interface is set up via kernel command line.  The primary
interface is supplied via kernel command line for AWS and VMware
variants.  This change also removes the accept-ra setting from the
`systemd-sysctl` config.  Allowing wicked to manage this setting
eliminates any chance of races between sysctl/kernel/wicked.

zmrow commented:
^ Fixed up the commits per @bcressey
